packages/pymedphys_dicom/tests/test_dicom_anonymise.py

```diff
@@ -19,33 +25,102 @@
-    assert elems_anon_manual == elems_anon_func
-
-    assert not is_anonymised_dataset(ds_anon_manual, ignore_private_tags=False)
-    assert not is_anonymised_dataset(ds_anon_func, ignore_private_tags=False)
-    assert is_anonymised_dataset(ds_anon_manual, ignore_private_tags=True)
-    assert is_anonymised_dataset(ds_anon_func, ignore_private_tags=True)
-
-    ds_anon_func_full = anonymise_dataset(ds, delete_private_tags=True)
-    assert is_anonymised_dataset(ds_anon_func_full, ignore_private_tags=False)
-
-
-def test_is_anonymised_file():
+VR_NON_ANONYMOUS_REPLACEMENT_VALUE_DICT = {
+    'AS': "1Y",
+    'CS': "SMITH",
+    'DA': "20190429",
+    'DT': "20190429000700.000000",
+    'LO': "Smith",
+    'PN': "Smith",
+    'SH': "Smith",
+    'SQ': [Dataset(), Dataset()],
+    'ST': "Smith",
+    'TM': "000700.000000",
+    'UI': "11111118"}
+
+
+def _get_non_anonymous_replacement_value(keyword):
+    """Get an appropriate dummy anonymisation value for a DICOM element
+    based on its value representation (VR)
+    """
+    vr = BASELINE_KEYWORD_VR_DICT[keyword]
+    return VR_NON_ANONYMOUS_REPLACEMENT_VALUE_DICT[vr]
+
+
+def test_anonymise_and_is_anonymised_dataset():
+
+    # Create dict with one instance of every identifying keyword and
+    # run basic anonymisation tests
+    non_anon_dict = dict.fromkeys(IDENTIFYING_KEYWORDS)
+
+    for key in non_anon_dict:
+        non_anon_dict[key] = _get_non_anonymous_replacement_value(key)
+
+    ds_non_anon = dicom_dataset_from_dict(non_anon_dict)
+    assert not is_anonymised_dataset(ds_non_anon)
+
+    ds_anon = anonymise_dataset(ds_non_anon)
+    assert is_anonymised_dataset(ds_anon)
+
+    # Test anonymisation (and check thereof) for each identifying
+    # element individually.
+    for elem in ds_anon.iterall():
+        ds_single_non_anon_value = deepcopy(ds_anon)
+        setattr(ds_single_non_anon_value,
+                elem.keyword,
+                _get_non_anonymous_replacement_value(elem.keyword))
+        assert not is_anonymised_dataset(ds_single_non_anon_value)
+
+        ds_single_anon = anonymise_dataset(ds_single_non_anon_value)
+        assert is_anonymised_dataset(ds_single_anon)
+
+    # Test correct handling of private tags
+    ds_anon.add(DataElement(0x0043102b, 'SS', [4, 4, 0, 0]))
+    assert not is_anonymised_dataset(ds_anon, ignore_private_tags=False)
+    assert is_anonymised_dataset(ds_anon, ignore_private_tags=True)
+
+    ds_anon.remove_private_tags()
+    assert is_anonymised_dataset(ds_anon, ignore_private_tags=False)
+
+    # Test blank anonymisation
+    assert not is_anonymised_dataset(ds_non_anon)  # Sanity check
+    ds_anon_blank = anonymise_dataset(ds_non_anon, replace_values=False)
+    assert is_anonymised_dataset(ds_anon_blank)
+
+    # Test handling of unknown tags by removing PatientName from
+    # baseline dict
+    patient_name_tag = tag_for_keyword('PatientName')
+    patient_name = BaselineDicomDictionary.pop(patient_name_tag)
```

Comment:
> How is this working here? ... I guess that does work doesn't it... It makes my skin crawl a little editing an imported dict like that... It makes me think we should use a tuple instead of a dict for the BaselineDicom... But maybe we are all "responsible adults" and letting people do this is just fine... ... It'll only be used when trying to break things...

Well, I guess we'll want an easy way to update the dict, which at the moment is a straight copy of `pydicom`'s dict, IIRC? I don't think it's *too* much of a risk, but I'm happy to consider making it immutable :)

> Maybe do the edit inside a try, finally pattern? So if this one fails we don't get really confused if we see other tests in other modules failing one day?

Yes, good idea.

Reply:
I think this can fall under the "we are all responsible adults" clause:

https://docs.python-guide.org/writing/style/#we-are-all-responsible-users

It can be left as is. It's not our job to know better than the user's of the code... If a piece of code wants to overwrite that library (maybe to write a test? Who knows?) We shouldn't stop them, no matter how much it makes my skin crawl.